### PR TITLE
Add bare-bones WS server for autobahn testing

### PR DESCRIPTION
**Pull request by mlodato517**
**This PR**
Hopefully creates a minimal autobahn websocket server example from which to start testing and addressing issues for https://github.com/actix/actix-web/issues/1006.

I assume this example needs to be tweaked slightly to make a reasonable benchmark. From there I can run the autobahn test suite and document all the failures. Then we can pick them off as we see fit.

### CODE REVIEW

**Comment by robjtede** [comment]
This is a great start. Pretty handy to have this examples repo around for test server implementations.

I'd rather not require installing the client through pip and virtualenv but use the docker image they provide.

Being on macOS, I've had success running tests using this set up:

`config/fuzzingclient.json`

```json
{
  "outdir": "./reports/servers",
  "servers": [
    {
      "url": "ws://host.docker.internal:9001"
    }
  ],
  "cases": ["*"],
  "exclude-cases": [],
  "exclude-agent-cases": {}
}
```

```sh
docker run -it --rm \
    -v "${PWD}/config:/config" \
    -v "${PWD}/reports:/reports" \
    --network host \
    --name autobahn \
    crossbario/autobahn-testsuite \
    wstest \
    --spec /config/fuzzingclient.json \
    --mode fuzzingclient
```

though on Linux it is probably more appropriate to use `--network host` and a 127.0.0.1 as the URL.

**Comment by mlodato517** [comment]
> I'd rather not require installing the client through pip and virtualenv but use the docker image they provide.

This is a great point! It's what [the docs](https://github.com/crossbario/autobahn-testsuite#using-the-testsuite-docker-image) suggest anyway:
> The following recipe still works, but the new, recommended way is using a Docker toolchain image we provide.

I just have no Docker experience so I avoided it (which is silly).

But I guess is there any reason for me to specify _any_ of this in our README? Should I just link to their README and let anyone who wants to run the test follow the docs the way they prefer? I did just notice that their only docker example is for testing a client, not a server, so I do feel like your example is valuable though 🤔 Maybe it should be added to their README? What do you think?

**Comment by robjtede** [comment]
> is there any reason for me to specify any of this in our README?

I don't see why not. I'd even include the config and reports directories to make it instant for anyone to clone the repo and run the conformance tests without having to look through autobahn docs.

**Comment by mlodato517** [comment]
@robjtede Added [two commits](https://github.com/actix/examples/pull/331/files/6a285ef43a423822e758d0a123d5ba6243a5b4be..f7901d03a8459907364bb62b6f2f6796ad97394b) to hopefully incorporate what you were looking for. Let me know if that's heading in the right direction!

**Comment by robjtede** [comment]
Yep, I like this approach much better. One more tiny detail though, since the reports folder is not necessary we should track a `reports/.gitkeep` file so folks don't have to mkdir it.

**Comment by mlodato517** [comment]
> One more tiny detail though, since the reports folder is not necessary we should track a `reports/.gitkeep` file so folks don't have to mkdir it.

I almost did that but it seemed to me that `wstest` creates the directory if it doesn't exist - do you see an error instead?

**Comment by robjtede** [comment]
> do you see an error instead

No but I can't speak for behavior on Windows or Linux.

**Comment by mlodato517** [comment]
> No but I can't speak for behavior on Windows or Linux.

Safety first! Should I put a `.gitkeep` in `reports/servers` or should we assume `wstest` creates _that_ directory on all systems?

**Comment by robjtede** [comment]
We can be sure it'll write to a dir called reports but I'd feel better having a .gitkeep around since we have the config folder there already.

**Comment by mlodato517** [comment]
Should there be any `.gitignore` updates to ignore reports that get put there? Or do we want those source controlled? Or do we want to rely on people to not add them?

**Comment by robjtede** [comment]
Good idea. A gitignore in that folder set to ignore everything except the .gitkeep would be good too. Probably don't want these tracked ever.

**Comment by mlodato517** [comment]
> Good idea. A gitignore in that folder set to ignore everything except the .gitkeep would be good too. Probably don't want these tracked ever.

Hmmm, sorry to bother you but what's the best way to do that? Should it be:
```
// reports/.gitignore
*
!.gitignore
!.gitkeep
```
or
```
// websocket-autobahn/.gitignore
reports/*
!reports/.gitkeep
```
or some other thing?

**Comment by robjtede** [comment]
np, now that I think about it a gitignore inside the reports folder would serve the same purpose as .gitkeep. File should look then look like this in reports:

```
*
!.gitignore
```

File: websocket-autobahn/config/fuzzingclient.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "outdir": "./reports/servers",
+  "servers": [
+    {
+      "agent": "actix-web-actors",
+      "url": "ws://host.docker.internal:9001"
+    }
+  ],
+  "cases": ["*"],
+  "exclude-cases": [],
```
Comment (mlodato517) [inline websocket-autobahn/config/fuzzingclient.json:10]: Oh, I also meant to ask - are there any tests we should be excluding by default? Not sure if `actix` is already supporting/planning-on-supporting per-message deflate and whatever else

Comment (robjtede) [reply websocket-autobahn/config/fuzzingclient.json:10]: Might as well for now. Looks like it's `"12.*"` and `"13.*"`.

Comment (mlodato517) [reply websocket-autobahn/config/fuzzingclient.json]: 51201fa807600c3bb0f68c4f06c0dc08cd073f01

**Comment by robjtede** [comment]
Great work! I don't have the bandwidth right now to tackle the failures since I'm trying to get v3 in a ready state so if you feel like investigating we'd appreciate it 😄.

**Comment by mlodato517** [comment]
> so if you feel like investigating we'd appreciate it 😄.

First of all - thanks for all the help with this! Secondly, definitely! The whole point of this PR was so that I could create a list to make other PRs off of to fix the other issues :-). I'll run the suite and document the errors on https://github.com/actix/actix-web/issues/1006 and then probably do 1 PR per issue or whatever's most convenient. I just need some help figuring out how to do that. Basically I need help linking to a local crate so I can update, say `actix-web` and then test that it fixed the issue via `actix:examples` or whatever.

**Comment by robjtede** [comment]
First of all, no problem at all. I'm usually responsive on [Gitter]()https://gitter.im/actix/actix-web or the [unoffical Discord](https://discord.gg/2A9h75V) if you want to ask Qs outside of issues. Thanks for the offer to look into this stuff.

Should be able to switch out the dependency on this project with something like `actix-web[-actors] = { path = "../actix-web" }` and then edit the actix-web repo files without any other changes. (You may still run into the usual issues where a `cargo clean` or `rm Cargo.lock` fixes 99% of the time.)

**Pull request merged**